content/pytorch/concepts/distributed-data-parallelism/distributed-data-parallelism.md

```diff
@@ -0,0 +1,105 @@
+---
+Title: 'Distributed Data Parallelism'
+Description: 'Enables users to efficiently train models across multiple GPUs and machines.'
+Subjects:
+  - 'Data Science'
+  - 'Machine Learning'
+Tags:
+  - 'Data'
+  - 'Data Parallelism'
+  - 'PyTorch'
+CatalogContent:
+  - 'intro-to-py-torch-and-neural-networks'
+  - 'paths/build-a-machine-learning-model'
+---
+
+## Introduction to Distributed Data Parallelism
+
+Distributed Data Parallelism (DDP) in PyTorch is a module that enables users to efficiently train models across multiple GPUs and machines. By splitting the training process across multiple machines, DDP helps reduce training time and facilitates scaling to larger models and datasets.
+It achieves parallelism by splitting the input data into smaller chunks, processing them on different GPUs, and aggregating results for updates. Compared to `DataParallel`, DDP offers better performance and scalability by minimizing device communication overhead.
+
+### Key Features
+
+- High performance: Efficient communication using NVIDIA Collective Communications Library (NCCL) or Gloo backend (for Windows platforms).
+- Fault tolerance: Handles errors during distributed training.
+- Scalability: Suitable for multi-node, multi-GPU setups.
+
+## Syntax
+
+To use DDP, a distributed process group needs to be initialised and wrapped to a model with `torch.nn.parallel.DistributedDataParallel`.
+
+```pseudo
+import os
+import torch
+import torch.distributed as dist
+from torch.nn.parallel import DistributedDataParallel as DDP
+
+# Set up environment variables
+os.environ['MASTER_ADDR'] = 'localhost'
+os.environ['MASTER_PORT'] = '8000'
+
+# Initialise the process group
+dist.init_process_group(backend="nccl", init_method="env://", world_size=4, rank=0)
+
+# Wrap the model with DDP
+model = torch.nn.Linear(10, 10).cuda()
+model = DDP(model)
+```
+
+- `backend`: This is optional and by default, a `gloo` and `nccl` backend will be created.
```

Comment:
This can be better written like this -

```suggestion
- `backend` (Optional): Specifies the backend for distributed communication.
	- `nccl` (NVIDIA Collective Communications Library):
		- Optimized for NVIDIA GPUs.
		- Supports high-performance GPU communication using CUDA.
		- Preferred choice for GPU-based distributed training.
	- `gloo`:
		- A cross-platform backend.
		- Used for CPU-based training or on systems without GPUs.
```